log/term/colorlogger.go

```diff
@@ -0,0 +1,149 @@
+package term
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+)
+
+// Color is the abstract color, the zero value is the Default.
+type Color uint8
+
+const (
+	NoColor = Color(iota)
+	Red
+	Green
+	Yellow
+	Blue
+	Magenta
+	Cyan
+	White
+	Default
+
+	maxColor
+)
+
+var resetColorBytes = []byte("\x1b[0m")
+var fgColorBytes [][]byte
+var bgColorBytes [][]byte
+
+func init() {
+	for color := NoColor; color < maxColor; color++ {
+		fgColorBytes = append(fgColorBytes, []byte(fmt.Sprintf("\x1b[%dm", 30+color)))
+		bgColorBytes = append(bgColorBytes, []byte(fmt.Sprintf("\x1b[%dm", 40+color)))
+	}
+}
+
+type FgBgColor struct {
+	Fg, Bg Color
+}
+
+func (c FgBgColor) IsZero() bool {
+	return c.Fg == NoColor && c.Bg == NoColor
+}
+
+// NewColorLogger returns a log.Logger which writes colored logs to w. It
+// colors whole records based on the FgBgColor returned by the color function.
+// Log events are formatted by the Logger returned by newLogger.
+func NewColorLogger(w io.Writer, newLogger func(io.Writer) log.Logger, color func(keyvals ...interface{}) FgBgColor) log.Logger {
+	if color == nil {
+		panic("color func nil")
+	}
```

Comment:
I think explicit is better than implicit and that baking this decision into `ColorLogger` control away from the application developer.

For example, I plan to use `ColorLogger` with github.com/mattn/go-colorable, which provides the `NewColorableStdout` and `NewColorableStderr` functions. On Linux these function just return `os.Stdout` and `os.Stderr` respectively, but on Windows they return an `io.Writer` that wraps the appropriate stream and converts the color codes into function calls to change the text color in the Windows console. The Windows `io.Writer`s would not satisfy `IsTty`.

There may be other situations where `ColorLogger` is only *indirectly* writing to a TTY. So, I believe `ColorLogger` should do what it says, and let the application decide if it is needed.

Also, without the check for a TTY, `ColorLogger` is easier to test.